test/integration/buoyancy.cc

```diff
@@ -42,6 +42,121 @@ class BuoyancyTest : public InternalFixture<::testing::Test>
 {
 };
 
+/////////////////////////////////////////////////
+TEST_F(BuoyancyTest, IGN_UTILS_TEST_DISABLED_ON_WIN32(RestoringMoments))
+{
+  // This test checks if the restoring moments are correctly calculated accross
+  // both uniform and graded modes when the vehicle is fully submerged.
+  std::size_t iterations{10000};
+  std::vector<math::Pose3d> poses_uniform, poses_graded;
+  {
+    // Start server
+    ServerConfig serverConfig;
+    const auto sdfFile = common::joinPaths(std::string(PROJECT_SOURCE_PATH),
+      "/test/worlds/buoyancy_uniform_restoring_moments.sdf");
+    serverConfig.SetSdfFile(sdfFile);
+
+    test::Relay testSystem;
+    testSystem.OnPostUpdate([&](const gazebo::UpdateInfo &/*unused*/,
+                              const gazebo::EntityComponentManager &_ecm)
+    {
+      // Get Submarine
+      Entity submarine = _ecm.EntityByComponents(
+        components::Model(), components::Name("submarine"));
+
+      // Get pose
+      auto submarinePose = _ecm.Component<components::Pose>(submarine);
+      ASSERT_NE(submarinePose, nullptr);
+
+      poses_uniform.push_back(submarinePose->Data());
+    });
+
+    Server server(serverConfig);
+    EXPECT_FALSE(server.Running());
+    EXPECT_FALSE(*server.Running(0));
+
+    server.AddSystem(testSystem.systemPtr);
+    server.Run(true, iterations, false);
+  }
+
+  {
+    // Start server
+    ServerConfig serverConfig;
+    const auto sdfFile = common::joinPaths(std::string(PROJECT_SOURCE_PATH),
+      "/test/worlds/buoyancy_graded_restoring_moments.sdf");
+    serverConfig.SetSdfFile(sdfFile);
+
+    test::Relay testSystem;
+    testSystem.OnPostUpdate([&](const gazebo::UpdateInfo &/*unused*/,
+                              const gazebo::EntityComponentManager &_ecm)
+    {
+      // Get Submarine
+      Entity submarine = _ecm.EntityByComponents(
+        components::Model(), components::Name("submarine"));
+
+      // Get pose
+      auto submarinePose = _ecm.Component<components::Pose>(submarine);
+      ASSERT_NE(submarinePose, nullptr);
+
+      poses_graded.push_back(submarinePose->Data());
+    });
+
+    Server server(serverConfig);
+    EXPECT_FALSE(server.Running());
+    EXPECT_FALSE(*server.Running(0));
+
+    server.AddSystem(testSystem.systemPtr);
+    server.Run(true, iterations, false);
+  }
+
+  double max_uniform_roll{0}, max_graded_roll{0};
+  double min_uniform_roll{0}, min_graded_roll{0};
+
+  ASSERT_EQ(poses_graded.size(), poses_uniform.size());
+  ASSERT_EQ(poses_uniform.size(), iterations);
+  for (std::size_t i = 0; i <  poses_graded.size(); i++)
+  {
+    // The two modes should track a similar course when fully submerged
+    // Ignore roll for now as that changes a lot and the two sims are not
+    // perfectly in sync.
```

Comment:
Both worlds are using the same step size and run for the same number of iterations. What's missing for them to be in sync?

Reply:
I'm not very sure but they do seem to diverge over time. I guess minor floating point errors build up? Again, this is something I'm speculating and not sure abot...